### PR TITLE
Prometheus Install Modes

### PR DESCRIPTION
**Pull request by charlieegan3**
Work is being completed as part of https://github.com/jetstack/tarmak/issues/349

```release-note
Enable the setting of a Prometheus installation 'mode' in the tarmak yaml that controls which components are installed. Also adds a new mode that only installs blackbox exporters on the etcd nodes - leaving them to be monitored by another Prometheus installation.
```

### CODE REVIEW

**Comment by charlieegan3** [comment]
This is what I get running on my master nodes for each of the Prometheus modes:

### Mode: `ExternalExportersOnly`

```
[centos@charlie-cluster|master|ip-10-99-53-213 ~]$ systemctl | grep prom
# nothing
[centos@charlie-cluster|master|ip-10-99-53-213 ~]$ systemctl | grep exporter
# nothing
```

### Mode: `ExternalScrapeTargetsOnly`

```
[centos@charlie-cluster|master|ip-10-99-59-90 ~]$ systemctl | grep prom
  kubectl-apply-prometheus-config.service                                                                                                                      loaded active exited    kubectl apply prometheus-config
  kubectl-apply-prometheus-rules.service                                                                                                                       loaded active exited    kubectl apply prometheus-rules
  kubectl-apply-prometheus-server.service                                                                                                                      loaded active exited    kubectl apply prometheus-server
[centos@charlie-cluster|master|ip-10-99-59-90 ~]$ systemctl | grep exporter
# nothing
```
Prometheus Config:
```
global:
  scrape_interval: 15s
  scrape_timeout: 10s
  evaluation_interval: 30s
  external_labels:
    cluster: charlie-cluster
rule_files:
- /etc/prometheus-rules/*.yaml
scrape_configs:
- job_name: etcd-nodes-exporter
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  static_configs:
  - targets:
    - etcd-1.charlie-cluster.tarmak.local:9100
    - etcd-2.charlie-cluster.tarmak.local:9100
    - etcd-3.charlie-cluster.tarmak.local:9100
    labels:
      role: etcd
- job_name: etcd-k8s-main
  params:
    module:
    - etcd_k8s_main_proxy
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /probe
  scheme: http
  static_configs:
  - targets:
    - etcd-1.charlie-cluster.tarmak.local:9115
    - etcd-2.charlie-cluster.tarmak.local:9115
    - etcd-3.charlie-cluster.tarmak.local:9115
    labels:
      role: etcd
  relabel_configs:
  - separator: ;
    regex: (.*)
    target_label: __param_target
    replacement: https://127.0.0.1:2379/metrics
    action: replace
- job_name: etcd-k8s-events
  params:
    module:
    - etcd_k8s_events_proxy
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /probe
  scheme: http
  static_configs:
  - targets:
    - etcd-1.charlie-cluster.tarmak.local:9115
    - etcd-2.charlie-cluster.tarmak.local:9115
    - etcd-3.charlie-cluster.tarmak.local:9115
    labels:
      role: etcd
  relabel_configs:
  - separator: ;
    regex: (.*)
    target_label: __param_target
    replacement: https://127.0.0.1:2369/metrics
    action: replace
- job_name: etcd-overlay
  params:
    module:
    - etcd_overlay_proxy
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /probe
  scheme: http
  static_configs:
  - targets:
    - etcd-1.charlie-cluster.tarmak.local:9115
    - etcd-2.charlie-cluster.tarmak.local:9115
    - etcd-3.charlie-cluster.tarmak.local:9115
    labels:
      role: etcd
  relabel_configs:
  - separator: ;
    regex: (.*)
    target_label: __param_target
    replacement: https://127.0.0.1:2359/metrics
    action: replace
```

### Mode: `Full`

```
[centos@charlie-cluster|master|ip-10-99-38-231 ~]$ systemctl | grep prom
  kubectl-apply-prometheus-config.service                                                                                                      loaded active exited    kubectl apply prometheus-config
  kubectl-apply-prometheus-rules.service                                                                                                       loaded active exited    kubectl apply prometheus-rules
  kubectl-apply-prometheus-server.service                                                                                                      loaded active exited    kubectl apply prometheus-server
[centos@charlie-cluster|master|ip-10-99-38-231 ~]$ systemctl | grep exporter
  kubectl-apply-blackbox-exporter.service                                                                                                      loaded active exited    kubectl apply blackbox-exporter
  kubectl-apply-node-exporter.service                                                                                                          loaded active exited    kubectl apply node-exporter
```

Prometheus config:
```
global:
  scrape_interval: 15s
  scrape_timeout: 10s
  evaluation_interval: 30s
  external_labels:
    cluster: charlie-cluster
rule_files:
- /etc/prometheus-rules/*.yaml
scrape_configs:
- job_name: kubernetes-apiservers
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: https
  kubernetes_sd_configs:
  - api_server: null
    role: endpoints
    namespaces:
      names: []
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    insecure_skip_verify: false
  relabel_configs:
  - source_labels: [__meta_kubernetes_namespace, __meta_kubernetes_service_name, __meta_kubernetes_endpoint_port_name]
    separator: ;
    regex: default;kubernetes;https
    replacement: $1
    action: keep
- job_name: kubernetes-nodes
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: https
  kubernetes_sd_configs:
  - api_server: null
    role: node
    namespaces:
      names: []
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    insecure_skip_verify: false
  relabel_configs:
  - separator: ;
    regex: __meta_kubernetes_node_label_(.+)
    replacement: $1
    action: labelmap
  - separator: ;
    regex: (.*)
    target_label: __address__
    replacement: kubernetes.default.svc:443
    action: replace
  - source_labels: [__meta_kubernetes_node_name]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: /api/v1/nodes/${1}/proxy/metrics
    action: replace
- job_name: kubernetes-nodes-cadvisor
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: https
  kubernetes_sd_configs:
  - api_server: null
    role: node
    namespaces:
      names: []
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    insecure_skip_verify: false
  relabel_configs:
  - separator: ;
    regex: __meta_kubernetes_node_label_(.+)
    replacement: $1
    action: labelmap
  - separator: ;
    regex: (.*)
    target_label: __address__
    replacement: kubernetes.default.svc:443
    action: replace
  - source_labels: [__meta_kubernetes_node_name]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: /api/v1/nodes/${1}/proxy/metrics/cadvisor
    action: replace
- job_name: kubernetes-nodes-exporter
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: https
  kubernetes_sd_configs:
  - api_server: null
    role: node
    namespaces:
      names: []
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    insecure_skip_verify: false
  relabel_configs:
  - separator: ;
    regex: __meta_kubernetes_node_label_(.+)
    replacement: $1
    action: labelmap
  - separator: ;
    regex: (.*)
    target_label: __address__
    replacement: kubernetes.default.svc:443
    action: replace
  - source_labels: [__meta_kubernetes_node_name]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: /api/v1/nodes/${1}:9100/proxy/metrics
    action: replace
- job_name: etcd-nodes-exporter
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  static_configs:
  - targets:
    - etcd-1.charlie-cluster.tarmak.local:9100
    - etcd-2.charlie-cluster.tarmak.local:9100
    - etcd-3.charlie-cluster.tarmak.local:9100
    labels:
      role: etcd
- job_name: etcd-k8s-main
  params:
    module:
    - etcd_k8s_main_proxy
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /probe
  scheme: http
  static_configs:
  - targets:
    - etcd-1.charlie-cluster.tarmak.local:9115
    - etcd-2.charlie-cluster.tarmak.local:9115
    - etcd-3.charlie-cluster.tarmak.local:9115
    labels:
      role: etcd
  relabel_configs:
  - separator: ;
    regex: (.*)
    target_label: __param_target
    replacement: https://127.0.0.1:2379/metrics
    action: replace
- job_name: etcd-k8s-events
  params:
    module:
    - etcd_k8s_events_proxy
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /probe
  scheme: http
  static_configs:
  - targets:
    - etcd-1.charlie-cluster.tarmak.local:9115
    - etcd-2.charlie-cluster.tarmak.local:9115
    - etcd-3.charlie-cluster.tarmak.local:9115
    labels:
      role: etcd
  relabel_configs:
  - separator: ;
    regex: (.*)
    target_label: __param_target
    replacement: https://127.0.0.1:2369/metrics
    action: replace
- job_name: etcd-overlay
  params:
    module:
    - etcd_overlay_proxy
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /probe
  scheme: http
  static_configs:
  - targets:
    - etcd-1.charlie-cluster.tarmak.local:9115
    - etcd-2.charlie-cluster.tarmak.local:9115
    - etcd-3.charlie-cluster.tarmak.local:9115
    labels:
      role: etcd
  relabel_configs:
  - separator: ;
    regex: (.*)
    target_label: __param_target
    replacement: https://127.0.0.1:2359/metrics
    action: replace
- job_name: kubernetes-service-endpoints
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  kubernetes_sd_configs:
  - api_server: null
    role: endpoints
    namespaces:
      names: []
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scheme]
    separator: ;
    regex: (https?)
    target_label: __scheme__
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_path]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: $1
    action: replace
  - source_labels: [__address__, __meta_kubernetes_service_annotation_prometheus_io_port]
    separator: ;
    regex: (.+)(?::\d+);(\d+)
    target_label: __address__
    replacement: $1:$2
    action: replace
  - separator: ;
    regex: __meta_kubernetes_service_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: kubernetes_namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_name
    replacement: $1
    action: replace
- job_name: kubernetes-services
  params:
    module:
    - http_2xx
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /probe
  scheme: http
  kubernetes_sd_configs:
  - api_server: null
    role: service
    namespaces:
      names: []
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_probe]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__address__]
    separator: ;
    regex: (.*)
    target_label: __param_target
    replacement: $1
    action: replace
  - separator: ;
    regex: (.*)
    target_label: __address__
    replacement: blackbox-exporter
    action: replace
  - source_labels: [__param_target]
    separator: ;
    regex: (.*)
    target_label: instance
    replacement: $1
    action: replace
  - separator: ;
    regex: __meta_kubernetes_service_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_service_namespace]
    separator: ;
    regex: (.*)
    target_label: kubernetes_namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_name
    replacement: $1
    action: replace
- job_name: kubernetes-ingresses
  params:
    module:
    - http_2xx
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /probe
  scheme: http
  kubernetes_sd_configs:
  - api_server: null
    role: ingress
    namespaces:
      names: []
  relabel_configs:
  - source_labels: [__meta_kubernetes_ingress_annotation_prometheus_io_probe]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_ingress_scheme, __address__, __meta_kubernetes_ingress_path]
    separator: ;
    regex: (.+);(.+);(.+)
    target_label: __param_target
    replacement: ${1}://${2}${3}
    action: replace
  - separator: ;
    regex: (.*)
    target_label: __address__
    replacement: blackbox-exporter
    action: replace
  - source_labels: [__param_target]
    separator: ;
    regex: (.*)
    target_label: instance
    replacement: $1
    action: replace
  - separator: ;
    regex: __meta_kubernetes_ingress_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_ingress_namespace]
    separator: ;
    regex: (.*)
    target_label: kubernetes_namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_ingress_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_name
    replacement: $1
    action: replace
- job_name: kubernetes-pods
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  kubernetes_sd_configs:
  - api_server: null
    role: pod
    namespaces:
      names: []
  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: $1
    action: replace
  - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
    separator: ;
    regex: ([^:]+)(?::\d+)?;(\d+)
    target_label: __address__
    replacement: ${1}:${2}
    action: replace
  - separator: ;
    regex: __meta_kubernetes_pod_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: kubernetes_namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_pod_name
    replacement: $1
    action: replace
```

File: pkg/tarmak/cluster/cluster.go [hunk 1]
```diff
@@ -297,6 +305,33 @@ func (c *Cluster) validateAPIServer() (result error) {
...
 }
 
+func (c *Cluster) validatePrometheusMode() error {
+	var result error
+
+	allowedModes := []string{
+		clusterv1alpha1.PrometheusModeFull,
+		clusterv1alpha1.PrometheusModeExternalScrapeTargetsOnly,
+		clusterv1alpha1.PrometheusModeExternalExportersOnly,
+	}
```
Comment (wallrj) [inline pkg/tarmak/cluster/cluster.go]: Drive by code review...

`sets.NewString` might be useful here.
*  https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets/string.go

```
allowedModes := sets.NewString(
    clusterv1alpha1.PrometheusModeFull, 
    clusterv1alpha1.PrometheusModeExternalScrapeTargetsOnly,
    clusterv1alpha1.PrometheusModeExternalExportersOnly,
)

if ! allowedModes.Has(modeString) {
...
   fmt.Errorf("allowed modes: %s", allowedModes.List()
...
}
```

Comment (charlieegan3) [reply pkg/tarmak/cluster/cluster.go]: Thanks - this is much nicer. Should be fixed in c0c771d647a9ff967b10637fe18c4f4a342aaa4e

**Pull request merged**